rclcpp/include/rclcpp/subscription.hpp

```diff
@@ -266,89 +149,107 @@ class Subscription : public SubscriptionBase
     rcl_interfaces::msg::IntraProcessMessage & ipm,
     const rmw_message_info_t & message_info)
   {
-    if (!get_intra_process_message_callback_) {
+    if (!use_intra_process_) {
       // throw std::runtime_error(
       //   "handle_intra_process_message called before setup_intra_process");
       // TODO(wjwwood): for now, this could mean that intra process was just not enabled.
       // However, this can only really happen if this node has it disabled, but the other doesn't.
       return;
     }
-    MessageUniquePtr msg;
-    get_intra_process_message_callback_(
-      ipm.publisher_id,
-      ipm.message_sequence,
-      intra_process_subscription_id_,
-      msg);
-    if (!msg) {
-      // This either occurred because the publisher no longer exists or the
-      // message requested is no longer being stored.
-      // TODO(wjwwood): should we notify someone of this? log error, log warning?
-      return;
-    }
-    any_callback_.dispatch_intra_process(msg, message_info);
-  }
-
-  using GetMessageCallbackType =
-    std::function<void (uint64_t, uint64_t, uint64_t, MessageUniquePtr &)>;
-  using MatchesAnyPublishersCallbackType = std::function<bool (const rmw_gid_t *)>;
-
-  /// Implemenation detail.
-  // TODO(ivanpauno): This can be moved to the base class. No reason to be here.
-  // Also get_intra_process_message_callback_ and matches_any_intra_process_publishers_.
-  void setup_intra_process(
-    uint64_t intra_process_subscription_id,
-    GetMessageCallbackType get_message_callback,
-    MatchesAnyPublishersCallbackType matches_any_publisher_callback,
-    IntraProcessManagerWeakPtr weak_ipm,
-    const rcl_subscription_options_t & intra_process_options)
-  {
-    std::string intra_process_topic_name = std::string(get_topic_name()) + "/_intra";
-    rcl_ret_t ret = rcl_subscription_init(
-      intra_process_subscription_handle_.get(),
-      node_handle_.get(),
-      rclcpp::type_support::get_intra_process_message_msg_type_support(),
-      intra_process_topic_name.c_str(),
-      &intra_process_options);
-    if (ret != RCL_RET_OK) {
-      if (ret == RCL_RET_TOPIC_NAME_INVALID) {
-        auto rcl_node_handle = node_handle_.get();
-        // this will throw on any validation problem
-        rcl_reset_error();
-        expand_topic_or_service_name(
-          intra_process_topic_name,
-          rcl_node_get_name(rcl_node_handle),
-          rcl_node_get_namespace(rcl_node_handle));
+    if (any_callback_.use_take_shared_method()) {
+      ConstMessageSharedPtr msg;
+      take_intra_process_message(
+        ipm.publisher_id,
+        ipm.message_sequence,
+        intra_process_subscription_id_,
+        msg);
+      if (!msg) {
+        // This either occurred because the publisher no longer exists or the
+        // message requested is no longer being stored.
+        // TODO(wjwwood): should we notify someone of this? log error, log warning?
```

Comment:
Let's add a log warning here (now that we have a logging system, we didn't at the time) and see if anyone complains. I think this shouldn't happen normally.